### DISCUSSION TITLE
Resolvers change the content of date and time

### DISCUSSION THREAD

**Discussion by robross0606** [Urigo/graphql-scalars#835] [General]
When using `TimeResolver` or `DateResolver`, the actual values are modified.  For example, if I put in a time value of "00:00:30Z", what is actually passed through to mutation resolver code for storage and response is a full DateTime which includes the current date.  
```
Error: expect(received).toEqual(expected) // deep equality

- Expected  - 1
+ Received  + 1

  Object {
-   "time": "00:00:30Z",
+   "time": "2020-09-04T00:00:30.000Z",
  } on '00:00:30Z'
```

This is not functionally accurate and changes the fundamental semantic meaning of the values.

**Comment by ardatan** [comment]
@robross0606 Could you share a reproduction repo?

**Comment by robross0606** [comment]
The spot where this happens looks pretty clear to me:

```javascript
export const parseTime = (time: string): Date => {
  const currentDateString = new Date().toISOString();
  return new Date(
    currentDateString.substr(0, currentDateString.indexOf('T') + 1) + time,
  );
};
```

Here you append the passed in time to the current date.  Since Javascript date/time is relative to January 1st, 1970, a "free standing" time value should probably be relative to that date instead of the current date.  Or at least there should be an option to do this:

```javascript
export const parseTime = (time: string): Date => {
  const currentDateString = new Date().toISOString();
  return new Date('1970-01-01T' + timeString + 'Z');
};
```

**Comment by robross0606** [comment]
However, personally I think converting these to a DateTime at all is a mistake.  Validation of the value is one thing.  Fundamentally altering the data type and, hence, the actual stored value and semantic meaning is another.  Pushing an ISO Time value into a full Javascript DateTime object is a round peg in a square hole.

**Comment by ardatan** [comment]
But it is not clear to me actually. :) `Date`, `Time` and `DateTime` scalars parse the input which is mostly a string value into JS Date. Nothing is converted to `DateTime`. You can use `RegularExpression` factory function to create your own scalars based on your own regular expressions.
https://github.com/Urigo/graphql-scalars#using-the-regularexpression-scalar

**Comment by robross0606** [comment]
Okay, sorry, a Javascript `Date` object which is actually a full DateTime.  You're changing the data, not just validating.  You're pushing in the current date so you can convert it to a Javascript `Date` object.  This is purely a time value.  Allow it to pass thru as such.

**Comment by robross0606** [comment]
I've already decided that your library isn't meeting my needs.  It is fundamentally altering data where I'm only looking to validate that the data conforms to ISO representation.  This seems perfectly fine to you, so I'll be implementing my own custom scalars and not using the library at all.

**Comment by ardatan** [comment]
The data is not changed actually, the time value is still there but it is a JS Date instance because there is no `Time` in JS. When it's serialized back, it becomes `00:00:30Z`. Also validations are done using the regular expressions created based on ISO standards(if not, that means there is a bug). Maybe our docs have missing information about this workflow. 
We're always open for discussions, ideas, contributions and more to make all those better :)

**Comment by robross0606** [comment]
What ends up being passed _into_ the actual resolvers from input is fundamentally changed.  Just because you convert it back and forth on the fringe, doesn't mean what is being handled internally based on the input is doing that.  At the very least, your documentation is very much lacking in detail.  I don't see anything that indicates these are even converted to Javascript `Date` objects.  It would be up to the user to recognize that either through trial and error or examining your code.  Even though they've used `TimeResolver` on an input value, what is coming in on the input (after your scalar parses it) contains bogus data in the date portion of the Javascript `date` which must be ignored.  Here's a very simple example:

GraphQL:
```
type Mutation {
  "Tests input validation."
  testValidation(input: ValidatedInput!): String!
}

input ValidatedInput {
  date: Date
  time: Time
  datetime: DateTime
}
```
Resolver javascript code:
```javascript
module.exports = {
  Mutation: {
    testValidation: async (_, { input }) => {
      // `input.time` contains the current date which is fundamentally wrong and we have to know to ignore it
      const response = Arango.save(input) // Ooops we've just persisted incorrect data by saving a full date value the user never input
      return safeStringify(input) // And here we spit back what was input purely so we can run a unit test.
    }
  }
}
```

Because of this, the TimeResolver scalar isn't even idempotent.  If I submit the same value tomorrow, it becomes a different value than it was today.

**Comment by marklawlor** [comment]
@ardatan I think two new scalars which simply validate the input and do not turn it into a Date object would be ideal. Eg `DateString` and `TimeString` which keep the type as a string.

Trying to use the `Date` and `Time` scalars with [graphql-code-generator](https://graphql-code-generator.com/) (another of The Guild's open source projects) is a bit frustrating. If you type the scalars as `string | Date`, your client-side code will allow `Date` objects at compile time but this will cause runtime errors with most popular GraphQL clients. So your forced to use `string` and do a TypeGuard in every resolver, which isn't ideal.

While I understand interactions with `graphql-code-generator` may not be your primary concern (not sure how TheGuild operate), the type conversion is annoying for Typescript users who are trying to maintain correct types.

**Comment by ardatan** [comment]
@marklawlor Why don't you think another scalar `DateTimeString` because the same conversion applies there? There is no types on the client side other primitive ones(string, number and boolean).
I am ok to have extra scalars that returns `string` instead of `Date` inside resolvers. Any PRs are welcome!

**Comment by marklawlor** [comment]
@ardatan Most GraphQL clients serialise objects with JSON.stringify.  Javascript's Date object has an unusual quirk where `JSON.stringify(new Date()) === new Date().toISOString()`. This is definitely not intended behaviour by the GraphQL clients, but this quirk allows you to keep the type `Date` for `DateTime` without it breaking anything.

**Comment by ardatan** [comment]
@marklawlor I know but `new Date(dateString).toISOString() === dateString` is not always true. So we should consider `DateTimeString` together with `DateString` and `DateTimeString` because people like @robross0606 would like to keep the timezone etc in that case.

**Comment by marklawlor** [comment]
Ah yes, forgot that toISOString converts to UTC. Seems reasonable to include DateTimeString :+1:

**Comment by vespertilian** [comment]
:wave:

I am also keen on adding a Time string to this library but my requirement is slightly different. My issue with the current time implementation is that forces a timezone either 'Z' (UTC) or the offset. What I am capturing in my app is the 24-hour time string without an offset, as the user is entering it irrespective of timezone. 

I imaging it would just validate "HH:MM:SS". That the hour range is 1 - 23, minute and second range 1-59

Would you be open to a 24TimeString (or some similar name)? I feel like TimeString is probably better reserved for the ISO version, but this is a unique and valid use case for a scalar.

**Comment by chrskrchr** [comment]
@vespertilian - to solve a similar "dates and times without timezones" issue, my team introduced `LocalDate` and `LocalTime` scalars that are backed by string value but include validation that they're in the right format.  They're intended to mimic the Java classes of the same name ([LocalDate](https://docs.oracle.com/javase/8/docs/api/java/time/LocalDate.html) and [LocalTime](https://docs.oracle.com/javase/8/docs/api/java/time/LocalTime.html)).  

The fact that they aren't deserialized into a JS `Date` object on the backend is intentional and forces devs to think thoughtfully about how they're dealing with the values.

I'd be happy to submit a PR to this project with those scalars if it sounds like they'd solve your issue.

**Comment by vespertilian** [comment]
@chrskrchr Hey thanks. Looks kinda like what I want. 

I have also already created an implementation. I think I like your name better. Happy for either of us to do a PR. As long as others are happy with the content and naming. 

My implementation for reference:

```ts
import { GraphQLError, GraphQLScalarType, Kind } from "graphql";

const UL_TIME = /(0[0-9]|1[0-9]|2[0-3]):[0-5][0-9]:[0-5][0-9]$/

export const GraphQLULTime = new GraphQLScalarType({
  name: 'ULTime',
  description: `
    Unqualified local time also known as 24 hour time and military time.
    Format: HH:mm:ss with no offset or timezone information.
    Start at 00:00:00 and has a maximum values of 23:59:59
    `,
  serialize(value: unknown) {
    if (typeof value !== 'string') {
      throw new TypeError(`Value is not string: ${value}`);
    }
    if (!UL_TIME.test(value)) {
      throw new TypeError(`Value is not valid unqualified local time (24 hour time): ${value} should be formatted HH:mm:ss`)
    }

    return value
  },
  parseValue(value: unknown) {
    if (typeof value !== 'string') {
      throw new TypeError(`Value is not string: ${value}`);
    }
    if (!UL_TIME.test(value)) {
      throw new TypeError(`Value is not valid unqualified local time (24 hour time): ${value} should be formatted HH:mm:ss`)
    }

    return value
  },
  parseLiteral(ast) {
    if (ast.kind !== Kind.STRING) {
      throw new GraphQLError(
        `Can only validate strings as ULTime but got a: ${ast.kind}`,
      );
    }
    if (!UL_TIME.test(ast.value)) {
      throw new TypeError(`Value is not valid unqualified local time (24 hour time): ${ast.value} should be formatted HH:mm:ss`)
    }

    return ast.value
  }
})

```

Do these specs line up with what you were thinking? 

```
import { GraphQLULTime } from "./ul-time-string";
import { Kind } from "graphql";

const VALID_VALUES = [
  '23:59:59',
  '00:00:00',
  '01:01:01',
  '11:11:11',
  '00:00:01',
  '01:00:00',
  '01:01:00'
]

const INVALID_VALUES = [
  'randomstring',
  '24:00:00',
  '25:00:00',
  '29:00:00',
  '00:60:00',
  '00:00:60',
  '23:59:59Z',
  '00:00:00Z',
  '00:00:00+01:30',
  '10:00:11.003Z'
]

const NON_STRING_VALUES = [
  123,
  {foo: 'bar'},
  false
]

describe('GraphQLUlTime', () => {
  describe('serialization', () => {
    it('throws an error when a non string is passed', () => {
      NON_STRING_VALUES.forEach((value) => {
        expect(() => GraphQLULTime.serialize(value))
          .toThrowError(`Value is not string: ${value}`)
      });
    });

    it('throws an error when a invalid string is passed', () => {
      INVALID_VALUES.forEach((value) => {
        expect(() => GraphQLULTime.serialize(value))
          .toThrowError(`Value is not valid unqualified local time (24 hour time): ${value} should be formatted HH:mm:ss`)
      });
    });

    it('returns the value for all valid values', () => {
      VALID_VALUES.forEach((value) => {
        expect(GraphQLULTime.serialize(value))
          .toEqual(value)
      });
    });
  })

  describe('parseValue', () => {
    it('throws an error when a non string is passed', () => {
      NON_STRING_VALUES.forEach((value) => {
        expect(() => GraphQLULTime.parseValue(value))
          .toThrowError(`Value is not string: ${value}`)
      });
    });

    it('throws an error when a invalid string is passed', () => {
      INVALID_VALUES.forEach((value) => {
        expect(() => GraphQLULTime.parseValue(value))
          .toThrowError(`Value is not valid unqualified local time (24 hour time): ${value} should be formatted HH:mm:ss`)
      });
    });

    it('returns the value for all valid values', () => {
      VALID_VALUES.forEach((value) => {
        expect(GraphQLULTime.parseValue(value))
          .toEqual(value)
      });
    });
  })

  describe('parseLiteral', () => {
    it('throws an error when the kind is not a string', () => {
      expect(() =>
        GraphQLULTime.parseLiteral({ value: '123', kind: Kind.INT }, {}),
      ).toThrow(`Can only validate strings as ULTime but got a: ${Kind.INT}`);
    })

    it('throws an error when a invalid string is passed', () => {
      INVALID_VALUES.forEach((value) => {
        expect(() => GraphQLULTime.parseLiteral({value, kind: Kind.STRING}, {}))
          .toThrowError(`Value is not valid unqualified local time (24 hour time): ${value} should be formatted HH:mm:ss`)
      });
    });

    it('returns the value for all valid values', () => {
      VALID_VALUES.forEach((value) => {
        expect(GraphQLULTime.parseLiteral({value, kind: Kind.STRING}, {}))
          .toEqual(value)
      });
    });
  });
})

```

**Comment by marklawlor** [comment]
I like the name LocalTime :)

@vespertilian Bit of a nitpick, but be aware the often military time does not include the colon and  [can include timezone in western countries](https://en.wikipedia.org/wiki/List_of_military_time_zones). For example, 6:00 a.m. in zone UTC−5 is written "0600R" and spoken "oh six hundred Romeo"

**Comment by vespertilian** [comment]
@marklawlor, yeah sure I can remove the military time reference, I only added it when reading Wikipedia and looking for other names. If it's not accurate, lets remove it. Thanks.

**Comment by chrskrchr** [comment]
@vespertilian - hope I'm not stepping on your toes, but I went ahead and submitted a PR of our solution that also includes support for a `LocalDate` scalar:

https://github.com/Urigo/graphql-scalars/pull/533

Our implementations of `LocalTime` are very similar, with the only noticeable difference is that our `LocalTime` *does* allow `24:00` as a valid value.  My project needs to use the `LocalTime` scalar to represent the exclusive end time of a time block, and so we need to support `24:00` as a signal that a block ends on midnight at the end of a day. For what it's worth, the ISO 8601 spec also allowed `24:00` as a valid time for the same reason, at least up until a 2019 revision of the spec:

https://en.wikipedia.org/wiki/ISO_8601#Times

So, I don't think this is necessarily a crazy thing to do.  😄 

Let me know what you all think.

**Comment by vespertilian** [comment]
@chrskrchr, hey yeah submit a PR. No issues here :).

I don't think we should have the default for a TS / JS library include the 24th hour, maybe this scalar could be a function you import and you pass a parameter `localTime({24h: true})`. I assume most people using the library would be using TS / JS and the various date libraries in this eco-system (Moment, Luxon, Date-Fns), having 24 will cause all of these to fail if you pass that as part of a date-time string. Maybe your scalar should be called JavaLocalTime JavaLocalDate? 

Or the scalar could accept 24:00:00 and rewrite it to 00:00:00? That way whenever it goes in or comes out there would be no errors in frontend/backend JS TS libraries.

**Comment by chrskrchr** [comment]
PR submitted here: https://github.com/Urigo/graphql-scalars/pull/533

> I don't think we should have the default for a TS / JS library include the 24th hour

That's a fair point.  Both Moment and Luxon _do_ currently support `24:00` as a valid time string, but I wouldn't want anyone to be surprised by that behavior.  

Rather than making it a function-scalar, I'd rather introduce a separate scalar called `LocalEndTime` that's identical to `LocalTime` but allows `24:00` and is intended to be used as an exclusive end time for a block of time.  I'd definitely prefer that over rewriting the value (which could be more surprising to client than allowing `24:00` as a value in the first place 😄)

**Comment by vespertilian** [comment]
@chrskrchr local end time sounds good.

**Comment by chrskrchr** [comment]
> local end time sounds good.

Great - I've split the time scalar into `LocalTime` and `LocalEndTime` where only the latter supports `24:00` as a valid value.

https://github.com/Urigo/graphql-scalars/pull/533

**Comment by vespertilian** [comment]
@chrskrchr 

We never actually created the DateTimeISOString did we? 

I think I want that now. I want the date validated against the ISO spec but I don't want it transformed into a Date object as I will just have to transform it back to write it to the DB. 

@ardatan Are you good for a PR with this?

**Comment by chrskrchr** [comment]
> We never actually created the DateTimeISOString did we?

Correct.

While the `LocalTime`, `LocalEndTime` and `LocalDate` scalars have the behavior you describe where they AREN'T transformed into a date object (they're left as strings), the strings can only hold times and dates that have no associated TZ.  We'd need a separate scalar for strings whose value can be any ISO8601 value.

**Comment by ardatan** [comment]
I moved this issue to Discussions.
